### PR TITLE
Run Razor LSP on host only

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes https://github.com/dotnet/roslyn/issues/49329

The Razor LSP server is being activated at the wrong time, which means it gets requests for documents it doesn't know about (eg, plain C# documents in a non-Razor project), and hence fails some contract checks.

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs [hunk 1]
```diff
@@ -27,6 +27,7 @@ namespace Microsoft.CodeAnalysis.ExternalAccess.Razor.Lsp
     /// </summary>
     [ContentType(ContentTypeNames.CSharpContentType)]
     [ClientName(ClientName)]
+    [RunOnContext(RunningContext.RunOnHost)]
```
Comment (davidwengier) [inline src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs:30]: This is the fix, the rest is in support of this line compiling successfully

**Comment by davidwengier** [comment]
I still need to test this properly but I don't have a build of VS with the new client package yet. It has been inserted though, in https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/250068

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs:30]: can you doc what this means?  it is greek to me.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs:30]: also, why is it not on other impls of AbstractInProcLanguageClient?  And can you doc that as well?

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs [hunk 2]
```diff
@@ -25,8 +25,16 @@ namespace Microsoft.CodeAnalysis.ExternalAccess.Razor.Lsp
...
+    /// <remarks>
+    /// This specifies RunOnHost because in LiveShare we don't want this to activate on the guest instance
+    /// because LiveShare drops the ClientName when it mirrors guest clients, so this client ends up being
+    /// activated solely by its content type, which means it receives requests for normal .cs and .vb files
+    /// even for non-razor projects, which then of course fails because it gets text sync info for documents
+    /// it doesn't know about.
+    /// </remarks>
     [ContentType(ContentTypeNames.CSharpContentType)]
     [ClientName(ClientName)]
+    [RunOnContext(RunningContext.RunOnHost)]
```
Comment (NTaylorMullen) [inline src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs:37]: This should be inferred. I'm really really surprised you have to add this. I have a suspicion that something else may have gone wrong here in our dependencies. I'd highly suggest reaching out to Miguel Lopez to validate that you need this.

Comment (davidwengier) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs:37]: Inferred from what? The presence of client name?

I still have been unable to validate this unfortunately, so I'm not banking on this just yet :)

**Comment by davidwengier** [comment]
Got a val build of this and verified that local razor works, and LiveShare non-razor doesn't error out 🎉

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs (hunk 2 above)
Comment (NTaylorMullen) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs:37]: "Inferred" may have been the wrong word choice. This should be the default for language clients

Comment (dibarbet) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs:37]: @NTaylorMullen I think RunOnHost means that the server is not mirrored, which is not the default (by default the server is mirrored to the guest)
https://devdiv.visualstudio.com/DevDiv/_git/VSLanguageServerClient?path=%2Fsrc%2Fproduct%2FLanguageServer.Client.Definition%2FRunOnContextAttribute.cs&version=GBdevelop&line=16&lineEnd=17&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs [hunk 3]
```diff
@@ -25,8 +25,16 @@ namespace Microsoft.CodeAnalysis.ExternalAccess.Razor.Lsp
     /// TODO - This can be removed once C# is using LSP for diagnostics.
     /// https://github.com/dotnet/roslyn/issues/42630
     /// </summary>
+    /// <remarks>
+    /// This specifies RunOnHost because in LiveShare we don't want this to activate on the guest instance
+    /// because LiveShare drops the ClientName when it mirrors guest clients, so this client ends up being
+    /// activated solely by its content type, which means it receives requests for normal .cs and .vb files
+    /// even for non-razor projects, which then of course fails because it gets text sync info for documents
```
Comment (dibarbet) [inline src/VisualStudio/Core/Def/Implementation/LanguageClient/RazorInProcLanguageClient.cs:32]: That's a long sentence!

Might be good to add a bit saying that even without text sync (and even if client name was mirrored) we don't really want to mirror any requests from the guest to this LSP server - the generated razor text buffers only exist on the host so only the host LSP client should be making requests to it (via the razor LSP server).

**Review by dibarbet — APPROVED** [review]

**Review by milopezc — APPROVED** [review]

**Review by NTaylorMullen — APPROVED** [review]

**Pull request merged**